### PR TITLE
Port mini vMac guest OS screen ROM patch.

### PR DESCRIPTION
**Pull request by rickyzhang82**
- Added Mac SE 24 bit ROM screen patch from mini vMac.
  The guest OS resolution in System 6 can be defined
  at an arbitrary combination.
- Mac frame buffer base mapping is moved out from guest RAM.
  This creates more leg room for the host frame buffer.
- Enable System 6 vscreen support in SDL2. Passed build and test.
- Enable System 6 vscreen support in SDL1. Passed build and test.
- Eanble System 6 vscreen support in XWidnow. Passed build and test.

See my [screen cast demo here](https://youtu.be/aXzM8t_u3zI)

Travis is great to catch my first PR. It detected that I haven't patched for SDL1 or XWindow. 

See [Travis indicated the build is green](https://travis-ci.org/github/rickyzhang82/macemu/builds/708523844)

I personally tested all 3 combinations and make sure this PR works.

Signed-off-by: Ricky Zhang <rickyzhang@gmail.com>

### CODE REVIEW

File: BasiliskII/src/SDL/video_sdl.cpp [hunk 1]
```diff
@@ -1025,6 +1033,11 @@ bool VideoInit(bool classic)
 	else if (default_height > sdl_display_height())
 		default_height = sdl_display_height();
 
+	// for classic Mac, make sure the display width is 8
```
Comment (asvitkine) [inline BasiliskII/src/SDL/video_sdl.cpp:1036]: Nit: "is divisible by 8" or "display width % 8 == 0"

File: BasiliskII/src/SDL/video_sdl.cpp [hunk 2]
```diff
@@ -1025,6 +1033,11 @@ bool VideoInit(bool classic)
...
+	if (classic) {
+		default_width = (default_width << 3) >> 3;
```
Comment (asvitkine) [inline BasiliskII/src/SDL/video_sdl.cpp:1038]: Nit: (default_width / 8) * 8

No reason to use shifts since compilers will generate the same instructions anyway.

File: BasiliskII/src/SDL/video_sdl2.cpp [hunk 3]
```diff
@@ -1398,6 +1406,10 @@ bool VideoInit(bool classic)
 	else if (default_height > sdl_display_height())
 		default_height = sdl_display_height();
 
+	// for classic Mac, make sure the display width is 8
```
Comment (asvitkine) [inline BasiliskII/src/SDL/video_sdl2.cpp:1409]: Same comments as in the other file.

File: BasiliskII/src/Unix/video_x.cpp [hunk 4]
```diff
@@ -1763,13 +1773,17 @@ bool VideoInit(bool classic)
...
-		mode_str = PrefsFindString("screen");
+	mode_str = PrefsFindString("screen");
 
 	// Determine display type and default dimensions
-	int default_width = 512, default_height = 384;
+	int default_width, default_height;
+	if (classic) {
+		default_width = 512;
+		default_height = 342;
+	}	else {
```
Comment (asvitkine) [inline BasiliskII/src/Unix/video_x.cpp:1783]: Nit: Weird whitespace.

File: BasiliskII/src/Unix/video_x.cpp [hunk 5]
```diff
@@ -1794,6 +1808,11 @@ bool VideoInit(bool classic)
 	else if (default_height > DisplayHeight(x_display, screen))
 		default_height = DisplayHeight(x_display, screen);
 
+	// for classic Mac, make sure the display width is 8
+	if (classic) {
+		default_width = (default_width << 3) >> 3;
+	}
```
Comment (asvitkine) [inline BasiliskII/src/Unix/video_x.cpp:1814]: Hmm, can you refactor this logic to a helper function that can be shared by all three files? Otherwise, duplicating this logic is pretty error prone.

File: BasiliskII/src/rom_patches.cpp [hunk 6]
```diff
@@ -1008,6 +1008,179 @@ static bool patch_rom_classic(void)
...
+
+	wp = (uint16 *)(ROMBaseHost + 0x8cc);
+	*wp = htons(MacScreenHeight);
+	wp = (uint16 *)(ROMBaseHost + 0x8ce);
+	*wp = htons(MacScreenWidth);
+
+	// blink floppy, disk icon
+	lp = (uint32 *)(ROMBaseHost + 0xf4c);
+	*lp = htonl(MacFrameBaseMac24Bit
+							+ (((MacScreenHeight / 4) * 2 - 25) * MacScreenWidth
```
Comment (asvitkine) [inline BasiliskII/src/rom_patches.cpp:1064]: Nit: This formatting is pretty hard to follow. Can you format it differently?

Alternatively, perhaps you can make a helper function that can be re-used in the different places below? The parameters could then also be named according to their meaning.

File: BasiliskII/src/rom_patches.cpp [hunk 7]
```diff
@@ -1008,6 +1008,179 @@ static bool patch_rom_classic(void)
...
+		*patchwp++ = htons(0xc2c5); /* MulU      D5, D1 */
+		patchCodeBaseMac += 2;
+
+		*patchwp++ = htons(0xd3c1); /* AddA.L    D1, A1 */
+		patchCodeBaseMac += 2;
+
+		*patchwp++ = htons(0x4e75); /* RTS */
+		patchCodeBaseMac += 2;
+		patchlp = (uint32 *)patchwp; // keep in sync
+	} else
```
Comment (asvitkine) [inline BasiliskII/src/rom_patches.cpp:1170]: Nit: Weird formatting?

File: BasiliskII/src/rom_patches.cpp [hunk 8]
```diff
@@ -1008,6 +1008,179 @@ static bool patch_rom_classic(void)
 	*wp++ = htons(M68K_EMUL_OP_IRQ);
 	*wp++ = htons(0x4a80);		// tst.l	d0
 	*wp = htons(0x67f4);		// beq		0x402be2
+
+
+	// Patch guest screen for an arbitary resolution
+	// We are going to steal the ROM space that runs RAM test
+	uint32 patchCodeBaseMac; 	// a stateful variable to keep track of the stolen location of the patched code in guest memory.
```
Comment (asvitkine) [inline BasiliskII/src/rom_patches.cpp:1015]: Nit: Why not declare this variable on the line it's first assigned (line 1026)? Also, I find the comment would be more readable to be above the line.

File: BasiliskII/src/rom_patches.cpp [hunk 9]
```diff
@@ -1008,6 +1008,179 @@ static bool patch_rom_classic(void)
...
+	uint16* patchwp; 					// a stateful word pointer pointing to the patch code in host memory.
```
Comment (asvitkine) [inline BasiliskII/src/rom_patches.cpp:1016]: Ditto here, just merge with line 1031.

Same for the others.

File: BasiliskII/src/Unix/video_x.cpp (hunk 5 above)
Comment (rickyzhang82) [reply BasiliskII/src/Unix/video_x.cpp:1814]: There are a large chunk of quite unique logic among SDL1, SDL2 and XWindows have  to initialize the width and the height.

Also, the add_mode functions in XWindows are different from SDL1, SDL2.

File: BasiliskII/src/Unix/video_x.cpp (hunk 4 above)
Comment (rickyzhang82) [reply BasiliskII/src/Unix/video_x.cpp:1783]: video_x cpp

They are tabs. It is consistent with the rest of the file.

File: BasiliskII/src/rom_patches.cpp (hunk 7 above)
Comment (rickyzhang82) [reply BasiliskII/src/rom_patches.cpp:1170]: rom_patches cpp

They are tabs. It is consistent with the rest of the file.

File: BasiliskII/src/rom_patches.cpp (hunk 8 above)
Comment (rickyzhang82) [reply BasiliskII/src/rom_patches.cpp:1015]: I want to single out them to add comments.

I moved all comment (none assembly) above the line

**Comment by rickyzhang82** [comment]
@asvitkine 

Thanks for your prompt response. I think the code format is fine. I double check they are all tabs. No spaces.

I rejected your request of merging SDL1, SDL2, XWindows setting code. Because there are macro and unique codes in XWindows. But I will think about refactoring the common code between SDL1 and SDL2. There is no fun to write and maintain any duplicate code. Also, I believe there are dead code in SDL2 which requires some cleaning. But it will not be in this PR.

File: BasiliskII/src/rom_patches.cpp (hunk 7 above)
Comment (rickyzhang82) [reply BasiliskII/src/rom_patches.cpp]: I got you. It is if/else. Done

File: BasiliskII/src/rom_patches.cpp (hunk 6 above)
Comment (rickyzhang82) [reply BasiliskII/src/rom_patches.cpp:1075]: rom_patches cpp

They are all tabs. See the image "->".

In Qt Creator, I set the visualization two spaces equals to one tab. Thus, you see the alignment. Unless you want to convert the tab into spaces, I don't see any better solution.

Comment (rickyzhang82) [reply BasiliskII/src/rom_patches.cpp]: Given the fact that we have wider screen, I just unwrapped the formulas. 
Done.

File: BasiliskII/src/Unix/video_x.cpp (hunk 4 above)
Comment (rickyzhang82) [reply BasiliskII/src/Unix/video_x.cpp]: I got it. It is hidden in my editor.

**Comment by rickyzhang82** [comment]
@asvitkine

Thanks for your prompt response. After carefully reviewing in vim, I found what you are after in the comment style. It is too bad that Qt Creator doesn't visualize it properly.

I rejected your request of merging SDL1, SDL2, XWindows setting code. Because there are macro and unique codes in XWindows. But I will think about refactoring the common code between SDL1 and SDL2. There is no fun to write and maintain any duplicate code. Also, I believe there are dead code in SDL2 which requires some cleaning. But it will not be in this PR.
